maven_iuvs/echelle.py

```diff
@@ -825,3 +966,1289 @@ def find_files_with_geometry(file_index):
+    DN_D_guess = sp.integrate.simpson(spectrum[D_a:D_b], x=wavelengths[D_a:D_b])
+
+    # central wavelength initial guess - go with the canonical value. There is no need to return a guess for D
+    # because it will be calculated as a constant offset from the H central line, per advice from Mike Stevens.
+    lambda_H_lya_guess = 121.567
+
+    # Background initial guess: assume a form y = mx + b. If m = 0, assume a constant offset.
+    bg_m_guess = 0
+    bg_b_guess = np.median(spectrum)
+
+    return [DN_H_guess, DN_D_guess, lambda_H_lya_guess, bg_m_guess, bg_b_guess]
+
+# Cosmic ray and hot pixel removal -------------------------------------------
+
+
+def get_lya_mask(datacube, light_fits):
+    """
+    Construct a mask for the Lyman alpha region to be used when cleaning up data. 
+    This is not a particularly good approach and tends to ruin the spectra shape, so this should probably be retired.
+    This also isn't currently used because the hot pixel routine is working ok and not deleting real emissions.
+
+    Parameters
+    ----------
+    datacube : array (3D)
+               Detector image array in (integrations, spatial bins, spectral bins)
+    light_fits : astropy.io.fits instance
+                 fits object representing the light observations.
+    Returns
+    ----------
+    mask : masked array
+    """
+    mask = np.zeros_like(datacube[0, :, :]) # only needs to be 2D.
+    si1, si2 = get_ech_slit_indices(light_fits) # slit
+    wi = np.asarray(np.where(np.logical_and(get_wavelengths(light_fits)>=121.53, get_wavelengths(light_fits)<=121.58)))[0]
+    mask[si1:si2+1, wi] = 1  # +1 because of the way python does indices.
+    return mask
+
+
+def remove_cosmic_rays(data, mask=None, Ns=2, std_or_mad="mad"): 
+    """
+    Removes cosmic rays from the detector image by stacking images and setting any pixel
+    which is outside the median ± Ns*sigma to the median value.
+
+    Parameters
+    ----------
+    data : Array (integrations, spatial bins, spectral bins)
+           All detector images for a given observation
+    themask : masked array
+              Mask shape to use for masking out lyman alpha.
+    Ns : int
+         number of sigma to constrain stacked-frame filtering
+    """
+    Nfr = data.shape[0]  # frames (integrations)
+    Nsp = data.shape[1]  # Spatial bins
+    Nw = data.shape[2]   # Wavelength bins
+
+    # Try winsorization before computing the median, to ensure it's robust
+    # Currently not implemented but should try again.
+    # sp.stats.mstats.winsorize(data, limits=(0, 0.01), inclusive=(True, True), inplace=True, nan_policy='raise')
+    
+    # this section looks across frames for any pixels that are outside the median+Ns*sigma. 
+    # pixels that are outside that range are set to the median value. 
+    medval = np.median(data, axis=0) # TODO: If this can be a vectorized form of median_high, it would match better with IDL pipeline.  
```

Comment:
could do `medhighval = np.min(np.ma.masked_array(data, mask=(data<medval)), axis=0).data`

Reply:
Works, despite appearing to be witchcraft

Reply:
Also made almost no difference to the fits (which is good)

Reply:
Fixed by: f0c60d192fd4cfebf083e2c2a9930d1f48e90bcf